File: web3-libs/mod3/playground/aula5/send_tx.py
from web3 import Web3

# Configuração do provider
RPC_URL = "http://127.0.0.1:8545"
web3 = Web3(Web3.HTTPProvider(RPC_URL))

# Configuração do signer
PRIVATE_KEY = "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80"
account = web3.eth.account.from_key(PRIVATE_KEY)


# Função para enviar uma transação
try:
    # Obtém o nonce do endereço
    nonce = web3.eth.get_transaction_count(account.address)

    # Define o destinatário e o valor da transação
    to_address = "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"
    value_in_ether = "1"

    # Obtém o preço atual do gás da rede
    gas_price = web3.eth.gas_price

    # Calcula o gás necessário para a transação
    gas_limit = web3.eth.estimate_gas(
        {
            "to": to_address,
            "from": account.address,
            "value": web3.to_wei(value_in_ether, "ether"),
        }
    )

    # Monta a transação
    transaction = {
        "to": to_address,
        "value": web3.to_wei(value_in_ether, "ether"),
        "gas": gas_limit,
        "gasPrice": gas_price,
        "nonce": nonce,
    }

    # Assina a transação
    signed_transaction = web3.eth.account.sign_transaction(transaction, PRIVATE_KEY)

    # Envia a transação
    receipt = web3.eth.send_raw_transaction(signed_transaction.raw_transaction)

    print("Transação enviada com sucesso! Hash:", receipt.hex())
except Exception as error:
    print("Erro ao enviar a transação:", error)

File: web3-libs/mod4/playground/aula5/sendTx.js
import { ethers } from "ethers";

// Configuração do provider
const RPC_URL = "http://127.0.0.1:8545";
const provider = new ethers.JsonRpcProvider(RPC_URL);

// Configuração do signer
const PRIVATE_KEY =
  "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80";
const wallet = new ethers.Wallet(PRIVATE_KEY).connect(provider);

// Função para enviar uma transação
async function sendTransaction(wallet) {
  try {
    // Obtém o nonce do endereço usando o provider
    const nonce = await provider.getTransactionCount(wallet.address);

    // Define o destinatário e o valor da transação
    const toAddress = "0x70997970C51812dc3A010C7d01b50e0d17dc79C8";
    const valueInEther = "1";

    // Obtém o preço atual do gás da rede
    const feeData = await provider.getFeeData(); // Obtém os dados da taxa
    const gasPrice = feeData.gasPrice; // Extraindo o gasPrice do feeData

    // Calcula o gás necessário para a transação
    const gasLimit = await provider.estimateGas({
      to: toAddress,
      from: wallet.address,
      value: ethers.parseEther(valueInEther),
    });

    // Monta a transação
    const transaction = {
      to: toAddress,
      value: ethers.parseEther(valueInEther),
      gasLimit: gasLimit,
      gasPrice: gasPrice,
      nonce: nonce,
    };

    // Envia a transação
    const txResponse = await wallet.sendTransaction(transaction);
    const receipt = await txResponse.wait();

    console.log(
      "Transação enviada com sucesso! Hash:",
      txResponse.hash // Alterado para acessar o hash diretamente do txResponse
    );
  } catch (error) {
    console.error("Erro ao enviar a transação:", error);
  }
}

// Chama a função para enviar a transação
sendTransaction(wallet);